drkafka/src/main/java/com/pinterest/doctorkafka/KafkaClusterManager.java

```diff
@@ -597,17 +596,54 @@ public UnderReplicatedReason getUnderReplicatedReason(String brokerHost,
 
   /**
    * Generate reassignment plan for dead brokers
+   * Different failure handling for locality-aware and non-locality-aware:
+   * - Locality-aware: will reassign topicPartitions that succeeded to get assigned to another
+   *   same-locality broker, but will send alert notifying that partitions that failed (due to
+   *   insufficient capacity on the same locality).
+   * - Non-locality-aware: current reassignment will fail (reassignments are all-or-none).
    */
   private Map<TopicPartition, Integer[]> generateReassignmentPlanForDeadBrokers(
       List<OutOfSyncReplica> outOfSyncReplicas) {
     Map<TopicPartition, Integer[]> replicasMap = new HashMap<>();
     boolean success = true;
+    boolean isLocalityAware = clusterConfig.enabledRackAwareness();
+
+    Map<String, PriorityQueue<KafkaBroker>> brokerQueueByLocality = null;
+    PriorityQueue<KafkaBroker> brokerQueue = null;
+    Map<OutOfSyncReplica, Map<Integer, String>> reassignmentToLocalityFailures = new HashMap<>();
+    if(isLocalityAware){
+      brokerQueueByLocality = kafkaCluster.getBrokerQueueByLocality();
+    } else {
+      brokerQueue = kafkaCluster.getBrokerQueue();
+    }
 
-    PriorityQueue<KafkaBroker> brokerQueue = kafkaCluster.getBrokerQueue();
     for (OutOfSyncReplica oosReplica : outOfSyncReplicas) {
-      Map<Integer, KafkaBroker> replacedNodes =
-          kafkaCluster.getAlternativeBrokers(brokerQueue, oosReplica);
+
+      double inBoundReq = replicaStatsManager.getMaxBytesIn(zkUrl, oosReplica.topicPartition);
+      double outBoundReq = replicaStatsManager.getMaxBytesOut(zkUrl, oosReplica.topicPartition);
+      int preferredBroker = oosReplica.replicaBrokers.get(0);
+
+      Map<Integer, KafkaBroker> replacedNodes;
+      replacedNodes = isLocalityAware
+                      ? kafkaCluster.getAlternativeBrokersByLocality(
+                          brokerQueueByLocality,
+                          oosReplica,
+                          inBoundReq,
+                          outBoundReq,
+                          preferredBroker,
+                          reassignmentToLocalityFailures
+                        )
+                      : kafkaCluster.getAlternativeBrokers(
+                          brokerQueue,
+                          oosReplica,
+                          inBoundReq,
+                          outBoundReq,
+                          preferredBroker
+                      );
       if (replacedNodes == null) {
+        if (isLocalityAware) {
+          continue;
```

Comment:
with this, will we exit the loop with `success == true` ? 

in that case, will we get an email alert if doctorkafka failes to find a locality aware assignment?

Reply:
@yuyang08 Yes, we will get a different email alert if the assignment partially failed and these failure will be captured in the hashmap `reassignmentToLocalityFailures`. L672-680 will then check if there are partial failures during locality aware reassignments and alert by email.

Reply:
Added a guard to set `success` to `false` if no locality-aware reassignments were made.